### PR TITLE
Add intro-to-functions-slides.md to foundersandcoders/coursebook/week-2

### PR DESCRIPTION
**Pull request by sima-qian**
This presentation, written as hackmd slides, introduces:
* functions as first-class objects
* callback functions
* higher-order functions
* the call stack
* synchronous vs. asynchronous code.

The presentation was given to FAC15 in Week 2 and, as suggested by @sofer, is being added as reference in case FAC15 wants to give it to FAC16.

Original hackmd can be found [here](https://hackmd.io/_CShxWvYTAmf3AQ0rJcGUg).

The presentation was given on Wednesday morning of Testing Week and was planned as a long lead-in to the [`morning-challenge-traffic-lights`](https://github.com/foundersandcoders/morning-challenge-traffic-lights). It overran, however, and ended up pushing the traffic light challenge back to Thursday morning.

**To be clear:** This is not being suggested as a permanent change to the curriculum at this point, I am just adding the slides as a reference in case somebody wants to make use of them in the future.

related #874

### CODE REVIEW

File: coursebook/week-2/intro-to-functions-slides.md [hunk 1]
```diff
@@ -0,0 +1,701 @@
...
+2.  **Q:** What do we expect to happen?
+3.  Run.
+4.  **Q:** Can someone talk us through what happened?
+5.  **Q:** How can we fix it?
+6.  **Solution:** `thirdFunction(function() {secondFunction(firstFunction)})`
+7.  Recap: to go more than one layer deep in callbacks, you need to wrap functions in anonymous functions.
+
+---
+
+## Can a callback function can take arguments?
```
Comment (sofer) [inline coursebook/week-2/intro-to-functions-slides.md]: repetition.

Comment (sima-qian) [reply coursebook/week-2/intro-to-functions-slides.md]: In my mind, this was purposefully explicit -- when the presentation was given students didn't immediately all shout out "yes, obviously!", so I think it was useful to ask and answer this specific question.

**Comment by arrested-developer** [comment]
Credit should be given to @shiryz (and @eliascodes) for the [readme](https://github.com/shiryz/call-it-back) this was developed from!

I love the way this was delivered by @sima-qian as a presentation, in particular that this session was an opportunity to unpick misconceptions with the students, and provided a lot of opportunities to do so (a side discussion about truthy and falsy values developed organically out of this too). The live REPLs were very important in demonstrating some of the behaviour of JS that's discussed.

Also I think a key aspect of what we've tried to do this week in FAC15 is to treat **callbacks as our friends 👬** - they're so fundamental to JS that it really isn't helpful to introduce them as this horrible thing (callback hell) that you'll hate but eventually you'll learn not to use.

Adding this in **before** the traffic light challenge eases a significant pain point in the curriculum. Traffic lights was still a high level of challenge, but an achievable challenge where most students at least could access it and the level of understanding at the end was solid.

**Comment by sima-qian** [comment]
> Credit should be given to @shiryz (and @eliascodes) for the readme this was developed from!

Absolutely! It's a fantastically structured readme that covers all of the bases 🙇

File: coursebook/week-2/intro-to-functions-slides.md (hunk 1 above)
Comment (shiryz) [reply coursebook/week-2/intro-to-functions-slides.md]: can is repeated though

Comment (sima-qian) [reply coursebook/week-2/intro-to-functions-slides.md]: Oh, I see! 😂

**Comment by eliascodes** [comment]
This is a reminder to myself: I'd like to merge this, but there's an issue with the call-stack related stuff that should be fixed first. I'm happy to do it myself, unless anyone else really wants to.

**Comment by sima-qian** [comment]
> This is a reminder to myself: I'd like to merge this, but there's an issue with the call-stack related stuff that should be fixed first. I'm happy to do it myself, unless anyone else really wants to.

I can fix it if you like. What's the issue?

**Comment by sima-qian** [comment]
@eliascodes Yep, I think that makes sense. I'll try and rewrite the callstack section this weekend 👍

**Comment by mattlub** [comment]
This looks really great!

The two things I noticed which could be clarified:

1. 
```js
foo1(foo2(foo3()));
```
This example coming after the callbacks and higher-order functions slide could be confusing, because we are not passing functions as arguments here, we are passing function results. (i.e. potential q's: what is `typeof foo3()`? what is typeof `foo2(foo3())`)

2. the returns in the async functions
```js
function asyncAddOne(number, callback) {
  setTimeout(function() {
    return callback(number + 1);
  }, 3000);
};
```
It is a common confusion I've seen students try to return values from async functions. There could be something on this- i.e. async functions typically pass results to a callback function instead of returning.

In this and the others cases the return could be removed. potential q: what does the `asyncAddOne` function return? (undefined)

**Comment by sima-qian** [comment]
Apparently time got away from me!

Changes have been made as recommended:
* call stack section corrected (@eliascodes)
* questions `typeof foo3()` and `typeof foo2(foo3())` added to the notes (@mattlub)
* `return` statements removed from async functions to avoid confusion (@mattlub)
* added a question about the return value of `asyncAddOne()` (@mattlub)

Is there anything else major that should be corrected or clarified before merging?

File: coursebook/week-2/intro-to-functions-slides.md [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+
+---
+
+# Calling back callbacks
+
+![](https://media.giphy.com/media/l1J9HX2NOtjEpIxJC/giphy.gif)
+
+---
+
+Recall that a `callback function` is simply a function that is passed as an argument to another function
```
Comment (oliverjam) [inline coursebook/week-2/intro-to-functions-slides.md]:
```suggestion
Recall that a `callback function` is a function that is passed as an argument to another function
```

Words like "simply" don't really help with understanding, and can be alienating for people who find this anything but simple

**Comment by oliverjam** [comment]
I love this, and you've clearly put in tons of effort ❤️ 

It will really help with Week 2, so I think we should try to get this merged asap.

My only slight concern is that this will require a mentor who understands this stuff very well and is willing to put in some time to get up to speed with all the notes and example code.

**Comment by sima-qian** [comment]
@oliverjam: Thanks! :) I've committed the change you suggested -- I recall now that someone in FAC14 brought up a similar point about language, so I really should have caught it myself! I've also changed it on the original hackmd document.

I'm aware @zurda is willing to give this presentation to FAC16, so I think it's worth merging before then if possible. Whether it gets taken up for FAC17 I guess will be a decision for the week 2 mentors coming from FAC16.

**Pull request merged**